docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,231 @@
+
+- [Remote Artifacts](https://github.com/shipwright-io/build/blob/master/docs/proposals/remote-artifacts.md)
+- [Build Inputs Overhaul](https://github.com/shipwright-io/build/pull/652)
+
+But to be pragmatic and avoid confusion, this EP proposes to do the following changes in the current state of the Build API.
+
+##### Modify the API
+
+- Introduce `spec.source.local`, which should host the desire usage of this feature. The value of `spec.source.local` requires to be an absolute path to the code in the user local machine. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L14-L16).
+
+- Ensure `spec.source.url` is not longer a mandatory field. This was done thinking on assets only hosted in `git`, which no longer holds true. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L19-L20).
+
+#### Modify the Runtime logic
+
+- Ensure that when `spec.source.local` is defined, we do not longer generate the Tekton Input PipelineResource. We do this today, to tell Tekton that we want to pull source from a git repository, which ends as a container that pulls it. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L184-L185) on future changes.
+
+- For the bundle mechanism, which is explained in the next section. We need to **prepend** a new step in our Task step definition, which will pull our local source code from a registry. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L171-L186). 
+
+  Important to notice, the image to pull will be a self-extracted image, therefore the `workingDir` container definition should be under `/workspace/source`, which is a well-known path in the Shipwright strategies, where source code is expected to be.
+
+_Note_: More changes code-wise might be needed, but that should be an implementation detail and does not belong in this EP description.
+_Note_: The above does not require a manual modification to any of the existing Shipwright strategies.
+
+
+### Part 2: Bundle Mechanism
+
+This is CLI specific. Any workflow using Shipwright that intends to provide support for local source code will require to re-use the Shipwright/CLI implementation or implement a standalone one.
+
+This proposed implementation can be one of many. In the future we might want to use a different storage service as needed.
+
+The Bundle mechanism has it´s origins from [Mink Bundles](https://github.com/mattmoor/mink/tree/master/pkg/bundles) and it´s based on the premise that if a user is pushing a container image to a container registry, then workflows could use the same approach to push the local source code, and pull it at a later stage during the building image mechanism.
+
+The Bundle mechanism executes the following:
+
+- [1] Takes the source code from a local directory.
```

Comment:
Just some random ideas here: Would it make sense to already consider that this could be _multiple_ local directories ? The reason I'm asking that because going over a registry via push-pull is per-se not a quick operation and the major use case is for a local development workflow, this round trip could be optimized by leveraging container image layering. I.e. one layer could contain the dependencies used by the application to build, one layer the static assets (resource files) and the top layer could then contain the application code (the JIB model for Java apps, but also very useful probably for runtimes with source dependency, like golang where one of the layers could be the `vendor/` dir). 

This would at least reduce the push-pull cycle, it probably does not help much that the self-extraction also costs time, especially when extracting a large dependency set (like a vendor dir). So I wonder, whether volume mounting of container directories could be an alternative to an extraction step ?

Reply:
@rhuss yes, we can consider this. As I mentioned somewhere before, this EP should not lock-in the proposed API changes  for the usage of container images as a way to provide source code at runtime, we can extend it with new ideas, like the one you are mentioning, with the aim to optimize performance. I would avoid putting this now, while this is just the basic API changes we need, and I dont really have strong use cases in mind for building such proposal, yet.